lib/ecto/type.ex

```diff
@@ -972,6 +972,23 @@ defmodule Ecto.Type do
     end
   end
 
+  @spec change(t, term, term) :: {:ok, term} | {:error, Ecto.Changeset.error()}
+  def change(type, term1, term2) do
```

Comment:
I don't think we need to add this function here. We can simply match and call change directly from `Ecto.Changeset`. :)

Reply:
Actually, ignore me. We need this function because we need to make it recursive to handle parameterized types inside arrays/maps. So that's what we should actually look into next.

Reply:
> I don't think we need to add this function here. We can simply match and call change directly from Ecto.Changeset. :)

I added it only for tests purposes, so I'm ok with moving it.

> Actually, ignore me. We need this function because we need to make it recursive to handle parameterized types inside arrays/maps. So that's what we should actually look into next.

So instead of `Ecto.Type` it should be in `Ecto.ParameterizedType`, right?

Reply:
No, sorry for the confusion. This function should actually be here. Here is what it would look like:

    def change(type, old_value, new_value) do
      change_fun(type).(old_value, new_value)
    end

    defp change_fun({:parameterized, mod, params}), do: &mod.change(&1, &2, params)
    defp change_fun(t) do
      fn old_value, new_value ->
        if equal?(type, old_value, new_value) do
          :skip
        else
          {:ok, value}
        end
      end
    end

And then in `put_change`, you will use the same code for both regular
types and paramterized ones:

    old_value = Map.get(data, key)

    case Ecto.Type.change(type, old_value, new_value) do
      {:ok, change} ->
        {Map.put(changes, key, change), errors, valid?}

      {:error, error} ->
        {message, opts} = Keyword.pop(error, :message, "cannot be changed")
        {changes, [{key, {message, opts}} | errors], false}

      :skip ->
        {Map.delete(changes, key), errors, valid?}
    end

Reply:
We also need to implement it for change_fun for arrays and maps, but it is a bit more complex. For example, if you have an array of enums, but one array is bigger than the other, what do you do? Similar for maps.